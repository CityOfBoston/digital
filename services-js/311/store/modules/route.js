// @flow
//
// Module for matching Redux store state to the Next router. Requires the
// withStoreRoute HOC to be mixed in to the page component.

type Query = {[key: string]: string};

type RoutePayload = {|
  pathname: string,
  query: Query,
  as: ?string,
|}

export type State = {
  pathname: ?string,
  query: ?Query,
  as: ?string,
}

export type Action = {| type: 'ROUTE_SET', payload: RoutePayload |};

// Causes navigation to the given Next pathname and query, using the "as"
// value for the window.location.
//
// Use this to trigger a route change from component code.
export const navigate = (pathname: string, query: Query, as: string): Action => ({
  type: 'ROUTE_SET',
  payload: { pathname, query, as },
});

// Updates the state based on a route change that happened. Should only be
// called by withStoreRoute.
export const setRoute = (pathname: string, query: Query): Action => ({
  type: 'ROUTE_SET',
  payload: { pathname, query, as: null },
});

const DEFAULT_STATE = {
  pathname: null,
  query: null,
  as: null,
};

export default function reducer(state: State = DEFAULT_STATE, action: Action): State {
  switch (action.type) {
    case 'ROUTE_SET':
      return {
        ...state,
        pathname: action.payload.pathname,
        query: action.payload.query,
        as: action.payload.as,
      };

    default:
      return state;
  }
}
