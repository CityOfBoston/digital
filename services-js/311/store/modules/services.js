// @flow

import { handle } from 'redux-pack';
import type { Deps } from '../';
import ServicesLoadGraphql from './graphql/ServicesLoad.graphql';
import ServicesLoadMetadataGraphql from './graphql/ServicesLoadMetadata.graphql';
import type { ServicesLoadMetadataQuery, ServicesLoadMetadataQueryVariables } from './graphql/schema.flow';

// TODO(finh): extract from ServicesLoadQuery when
// https://github.com/facebook/flow/issues/3379 is resolved
export type Service = {
  code: string,
  name: string,
  hasMetadata: boolean,
};

export type ServiceMetadata = $PropertyType<ServicesLoadMetadataQuery, 'serviceMetadata'>;

export type Action =
  {| type: 'SERVICES_LOAD', payload: any |} |
  {| type: 'SERVICES_LOAD_METADATA', payload: any, meta: { code: string } |};

export type State = {
  services: ?Service[],
  servicesError: ?Object,
  serviceMetadata: {[key: string]: ServiceMetadata | Error | null}
};

export const loadServices = () => ({ loopbackGraphql }: Deps) => ({
  type: 'SERVICES_LOAD',
  promise: loopbackGraphql(ServicesLoadGraphql),
});

export const loadServiceMetadata = (code: string) => ({ loopbackGraphql }: Deps) => ({
  type: 'SERVICES_LOAD_METADATA',
  promise: loopbackGraphql(ServicesLoadMetadataGraphql, ({
    code,
  }: ServicesLoadMetadataQueryVariables)),
  meta: {
    code,
  },
});

const DEFAULT_STATE = {
  services: null,
  servicesError: null,
  serviceMetadata: {},
};

export default function reducer(state: State = DEFAULT_STATE, action: Action): State {
  switch (action.type) {
    case 'SERVICES_LOAD': {
      const { payload } = action;
      return handle(state, action, {
        success: (s) => ({ ...s, services: payload.services, servicesError: null }),
        failure: (s) => ({ ...s, services: null, servicesError: payload }),
      });
    }

    case 'SERVICES_LOAD_METADATA': {
      const { payload, meta } = action;
      return handle(state, action, {
        success: (s) => ({ ...s, serviceMetadata: { ...s.serviceMetadata, [meta.code]: payload } }),
        failure: (s) => ({ ...s, serviceMetadata: { ...s.serviceMetadata, [meta.code]: payload } }),
      });
    }

    default:
      return state;
  }
}
