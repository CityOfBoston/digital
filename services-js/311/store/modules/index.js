// @flow
//
// Combines the exported reducers, State and Action types across all of the
// modules to provide single values to the store module.

import { combineReducers } from 'redux';
import type { Reducer } from 'redux';

import type { Deps, Dispatch } from '../';
import keys from './keys';
import services from './services';
import route from './route';
import request from './request';

import type { State as KeysState, Action as KeysAction } from './keys';
import type { State as ServicesState, Action as ServicesAction } from './services';
import type { State as RouteState, Action as RouteAction } from './route';
import type { State as RequestState, Action as RequestAction } from './request';

export type AppState = {|
  keys: KeysState,
  services: ServicesState,
  request: RequestState,
  route: RouteState,
|};

// Here we build up the Action type that matches the allowed inputs to our
// store’s dispatch function (in other words, the valid return values for our
// action creators.)
//
// This differs from the actions that the reducers can accept due to our
// store’s installed middleware.

// These are the Actions that the individual module reducers can handle
// natively.
type ReducableAction = KeysAction | RequestAction | RouteAction | ServicesAction;

// Action handled by redux-pack, which interprets a returned promise property
// and calls the reducer several times over its lifecycle.
type ReduxPackAction = {|
  type: string,
  promise: Promise<any>,
  meta?: Object,
|};

// Action that is handled by redux-thunk middleware, which calls the
// returned function with dispatch and getState.
type ThunkAction = (dispatch: Dispatch, getState: () => AppState) => any;

// Action handled by redux-inject, which takes a function and calls it with
// our injected dependencies. Its return value can be any of the above actions.
type InjectedAction = (Deps) => ThunkAction | ReduxPackAction | ReducableAction;

// The complete type that our dispatch accepts. Note that we don’t accept
// a ThunkAction here because any function passed to dispatch is handled by
// redux-inject. It’s only if a redux-inject function returns a function that
// redux-thunk gets called.
export type AppAction = InjectedAction | ReducableAction | ReduxPackAction;

export const reducer: Reducer<AppState, AppAction> = combineReducers({
  keys,
  services,
  route,
  request,
});
