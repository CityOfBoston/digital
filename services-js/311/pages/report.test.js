// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import Report from './report';

import { makeServerContext } from '../lib/test/make-context';
import inBrowser from '../lib/test/in-browser';
import getStore from '../store';
import { navigate } from '../store/modules/route';

jest.mock('../store/modules/services');

test('request flow', async () => {
  const ctx = makeServerContext('/report', { step: 'report' });
  const props = await Report.getInitialProps(ctx);

  await inBrowser(async () => {
    const component = renderer.create(<Report {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const store = getStore();
    await store.dispatch(navigate('/report', { step: 'contact' }, '/report/contact'));

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
