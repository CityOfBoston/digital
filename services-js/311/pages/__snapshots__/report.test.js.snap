exports[`test request flow 1`] = `
<div>
  <div
    className="nav ft-ll css-1avoji2">
    <a
      href="/"
      onClick={[Function]}>
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}>
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}>
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}>
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-saiqwd": "",
      }
    }>
    <img
      alt=""
      className={
        Object {
          "data-css-63oe3q": "",
        }
      }
      src="https://maps.googleapis.com/maps/api/staticmap?center=Boston%20Common&zoom=13&size=1200x800&maptype=roadmap&scale=2&key=FAKE_GOOGLE_API_KEY" />
  </div>
  <div
    className={
      Object {
        "data-css-1lz9nu": "",
      }
    }>
    <div
      className="sh">
      <h2
        className="sh-title">
        311: Boston City Services
      </h2>
    </div>
    <form
      action="/report/submit"
      className={
        Object {
          "data-css-1cqgl9p": "",
        }
      }
      method="POST">
      <input
        name="initialState"
        type="hidden"
        value="{\"request\":{\"description\":\"\",\"code\":null,\"firstName\":\"\",\"lastName\":\"\",\"email\":\"\",\"phone\":\"\",\"submitting\":false,\"submitError\":null}}" />
      <input
        name="actions"
        type="hidden"
        value="[{\"type\":\"REQUEST_SET_CODE\",\"payload\":\"code\"},{\"type\":\"REQUEST_SET_DESCRIPTION\",\"payload\":\"description\"}]" />
      <textarea
        defaultValue=""
        name="description"
        onInput={[Function]} />
      <ul>
        <li>
          <button
            name="code"
            onClick={[Function]}
            type="submit"
            value="needles">
            Needle Pickup
          </button>
        </li>
      </ul>
    </form>
  </div>
</div>
`;

exports[`test request flow 2`] = `
<div>
  <div
    className="nav ft-ll css-1avoji2">
    <a
      href="/"
      onClick={[Function]}>
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}>
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}>
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}>
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-saiqwd": "",
      }
    }>
    <img
      alt=""
      className={
        Object {
          "data-css-63oe3q": "",
        }
      }
      src="https://maps.googleapis.com/maps/api/staticmap?center=Boston%20Common&zoom=13&size=1200x800&maptype=roadmap&scale=2&key=FAKE_GOOGLE_API_KEY" />
  </div>
  <div
    className={
      Object {
        "data-css-1lz9nu": "",
      }
    }>
    <div>
      <h2>
        CONTACT FORM
      </h2>
      <p>
        (will not be shared with public; leave blank to submit anonymously)
      </p>
      <label
        className={
          Object {
            "data-css-572cbq": "",
          }
        }>
        <span>
          First Name
        </span>
        <input
          className={
            Object {
              "data-css-1etfc0l": "",
            }
          }
          name="firstName"
          onChange={[Function]}
          placeholder="First Name"
          value="" />
      </label>
      <label
        className={
          Object {
            "data-css-572cbq": "",
          }
        }>
        <span>
          Last Name
        </span>
        <input
          className={
            Object {
              "data-css-1etfc0l": "",
            }
          }
          name="lastName"
          onChange={[Function]}
          placeholder="Last Name"
          value="" />
      </label>
      <label
        className={
          Object {
            "data-css-572cbq": "",
          }
        }>
        <span>
          Email
        </span>
        <input
          className={
            Object {
              "data-css-1etfc0l": "",
            }
          }
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value="" />
      </label>
      <label
        className={
          Object {
            "data-css-572cbq": "",
          }
        }>
        <span>
          Phone
        </span>
        <input
          className={
            Object {
              "data-css-1etfc0l": "",
            }
          }
          name="phone"
          onChange={[Function]}
          placeholder="Phone"
          type="tel"
          value="" />
      </label>
      <button
        disabled={false}
        onClick={[Function]}
        type="submit">
        Submit
      </button>
    </div>
  </div>
</div>
`;
